wgengine/magicsock/magicsock_linux.go

```diff
@@ -0,0 +1,316 @@
+		bpf.LoadConstant{Dst: bpf.RegX, Val: ethHeaderSize + ipv6.HeaderLen},
+		bpf.Jump{Skip: 6}, // move past fragment checks
+
+		// Check if the "Next Header" field is the "Fragment" header
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: 0x2C, SkipTrue: 1, SkipFalse: 0},
+		bpf.RetConstant{Val: 0x0}, // skip packet
+
+		// If it's the fragment header, check whether the protocol is UDP
+		bpf.LoadAbsolute{Off: ethHeaderSize + ipv6.HeaderLen, Size: 1},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(ipproto.UDP), SkipTrue: 1, SkipFalse: 0},
+		bpf.RetConstant{Val: 0x0}, // skip packet
+
+		// Fragment header's "Next Header" indicates that we have a UDP
+		// packet immediately following; the offset is immediately
+		// following both headers.
+		bpf.LoadConstant{Dst: bpf.RegX, Val: ethHeaderSize + ipv6.HeaderLen + ipv6FragmentHeaderSize},
+
+		// If we get here, we have a UDP packet; compare with our magic
+		// number. Start by loading and comparing the first 4 bytes of
+		// the UDP packet.
+		bpf.LoadIndirect{Off: udpHeaderSize, Size: 4},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: magic1, SkipTrue: 0, SkipFalse: 3},
+
+		// Compare the next 2 bytes
+		bpf.LoadIndirect{Off: udpHeaderSize + 4, Size: 2},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(magic2), SkipTrue: 0, SkipFalse: 1},
+
+		// Accept the whole packet
+		bpf.RetConstant{Val: 0xFFFFFFFF},
+
+		// Skip the packet
+		bpf.RetConstant{Val: 0x0},
+	}
+)
+
+// listenDisco starts listening for disco packets using an AF_PACKET socket + a
+// BPF filter. This allows us to receive disco packets even without opening the
+// firewall; see issue 3824 for more detail.
+func (c *Conn) listenDisco() {
+	if debugDisableAF_PACKET {
+		return
+	}
+	c.listenDiscoProto(&c.bpfConn4, "IPv4", magicsockFilterV4)
+	c.listenDiscoProto(&c.bpfConn6, "IPv6", magicsockFilterV6)
+}
+
+func (c *Conn) listenDiscoProto(fdptr *int, name string, filter []bpf.Instruction) {
+	// We create our socket manually here, rather than using
+	// {ipv4,ipv6}.NewPacketConn, because they don't support SOCK_RAW.
+	fd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_RAW, int(htons(unix.ETH_P_ALL)))
+	if err != nil {
+		c.logf("magicsock: error creating raw socket for %s: %v", name, err)
+		return
+	}
+
+	*fdptr = fd
+	go func() {
+		err := c.listenPacketsWithFilter(fd, filter, c.handleDiscoPacket)
+		if err != nil {
+			c.logf("magicsock: error listening for %s disco packets with AF_PACKET: %v", name, err)
+		}
+	}()
+}
+
+func (c *Conn) closeListenDisco() {
+	if c.bpfConn4 != 0 {
+		unix.Close(c.bpfConn4)
+	}
+	if c.bpfConn6 != 0 {
+		unix.Close(c.bpfConn6)
+	}
+}
+
+// handleDiscoPacket is called by listenPacketsWithFilter when a packet
+// matching a BPF filter is received. It can be passed either an IPv4 or IPv6
+// packet and will handle either case.
+func (c *Conn) handleDiscoPacket(packet []byte, from unix.Sockaddr) error {
+	pkt := gopacket.NewPacket(packet, layers.LayerTypeEthernet, gopacket.NoCopy)
```

Comment:
Another benefit of using net.ListenPacket above, is that you can use net/packet to pull the packet apart here, and avoid the gopacket dependency just for dissecting ethernet.